### PR TITLE
scheduler-perf: run as integration tests

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This has two purposes:
- By running in the pre-merge pull-kubernetes-integration, changes that break scheduler_perf might be caught before they get merged.
- Eventually (see https://github.com/kubernetes/kubernetes/pull/116980), these integration tests will run with race detection enabled. This fills a gap in the test coverage for the kube-scheduler code because unit tests are often too simplistic to expose races and E2E testing runs without race detection.

#### Special notes for your reviewer:

Split into multiple stand-alone commits to simplify reviews, but probably none of those are worth merging by their own.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

**Review by mimani68 — APPROVED** [review]

**Comment by pohly** [comment]
> FAIL: TestScheduling/SchedulingPreferredPodAffinity/500Nodes

But there's no failure message :cry: 

/retest

**Comment by pohly** [comment]
Here's why the failure is not shown:
> clipping failure message in test case : TestScheduling/SchedulingPreferredPodAffinity/500Nodes

That's from `cmd/prune-junit-xml/prunexml.go`.

Verbosity is a bit high. The V(5) log message from graph_builder.go are too much. I'm not sure yet where that gets increased. The default should be -v2.

**Comment by pohly** [comment]
The verbosity problem was fixed by reconfiguring defaults for integration tests and the test failure should be fixed by disabling the sampling interval check. It's only relevant when we measure performance and is more likely to fail when many Go tests run in parallel.

**Comment by dims** [comment]
@Huang-Wei @ahg-g @alculquicondor can one of you please approve/lgtm?

**Comment by Huang-Wei** [comment]
@kerthcet could you take a look as you recently reviewed similar PRs in pert tests. Thanks!

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (chendave) [inline test/integration/scheduler_perf/scheduler_perf_test.go]: > Each benchmark test case runs with a fresh etcd instance. Therefore it is not
necessary to delete objects after a run.

yes, it's a fresh etcd instance, but I think it's always a good habit to do the cleanup after the test, this is also true for us to delete the object before removing the namespace entirely, it's not necessary but it's not something we should get rid of.

> A future unit test might reuse etcd, therefore cleanup is optional.

I don't think a future ut should reuse the leftover etcd instance.

File: test/integration/util/util.go
Comment (chendave) [inline test/integration/util/util.go:132]: I'd prefer to see there is some err or warning here to say that context is cancelled, or some error is occurred, unless this is already print to somewhere else.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (chendave) [inline test/integration/scheduler_perf/scheduler_perf_test.go:752]: it's actually integration test, and I am -1 for those integration test to share the same etcd.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (chendave) [inline test/integration/scheduler_perf/scheduler_perf_test.go:738]: Can we make the usage string a little different with the `perfSchedulingLabelFilter`, saying, this is integration test and the other is perf test?

File: test/integration/scheduler_perf/config/performance-config.yaml
Comment (chendave) [inline test/integration/scheduler_perf/config/performance-config.yaml:13]: What's the rationale for which test should be integration-test and others are not?

File: test/integration/util/util.go
Comment (pohly) [reply test/integration/util/util.go:132]: "Context cancelled" is normal and not an unexpected error. Therefore it doesn't need to be logged.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:752]: Agreed, "unit tests" is a leftover and needs to be removed. I don't agree that integration tests shouldn't share the same etcd. Most of them do, via https://github.com/kubernetes/kubernetes/blob/18d05b646d09b2971dc5400bc288062b0414e8cf/test/integration/framework/etcd.go#L186-L222

**Comment by chendave** [comment]
/assign

assign to myself, so I won't miss this one if this is still open.

File: test/integration/scheduler_perf/config/performance-config.yaml
Comment (chendave) [inline test/integration/scheduler_perf/config/performance-config.yaml:13]: nice writeup!

File: test/integration/util/util.go
Comment (alculquicondor) [reply test/integration/util/util.go:132]: Instead, we should probably be canceling the event handlers when the context is cancelled.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (alculquicondor) [reply test/integration/scheduler_perf/scheduler_perf_test.go:35]: Ah, this is just gomega, but still. Use cmp instead.

Comment (alculquicondor) [reply test/integration/scheduler_perf/scheduler_perf_test.go:35]: And you can use `k8s.io/apimachinery/pkg/util/wait`

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (kerthcet) [inline test/integration/scheduler_perf/scheduler_perf_test.go:934]: Maybe we can reuse `CleanupPods` in `test/integration/util/util.go`, or add a new function to reuse `CleanupPods`, like `CleanupPodsInNamespace(namespace string)`.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (kerthcet) [inline test/integration/scheduler_perf/scheduler_perf_test.go:945]: If `cleanup` is true, doesn't that mean **`namespaces` is empty** is enough? Or at least we can directly return when we want to append below, because that means the objects isn't empty.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (kerthcet) [inline test/integration/scheduler_perf/scheduler_perf_test.go:699]: Maybe next is we can setup the api-server first then only recreate the scheduler with each test. This differs from integration tests with benchmark test.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:35]: `util/wait` is awful for use in tests. All of the arguments from https://github.com/kubernetes/contributor-site/blob/bb1c87d320aacacc6ec5dd23d312fc59f93e3c6d/content/en/blog/2023/e2e-testing-best-practices.md#polling-and-timeouts also apply here.

I don't want to reimplement `gomega.Eventually`. If you insist, then I'll use `util/wait` and pretend that I don't know better.

File: test/integration/scheduler_perf/config/performance-config.yaml
Comment (pohly) [reply test/integration/scheduler_perf/config/performance-config.yaml:521]: They still take time, and pull-kubernetes-integration is already one of the longest running jobs for a PR.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:699]: I'm not sure what you are suggesting. Can you rephrase and explain what benefit it has?

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:945]: The purpose here is to provide a complete report of everything that hasn't been removed when removal doesn't finish in time. That is more accurate than reporting just one object, which might give the wrong impression.

Not sure whether that addresses your concern?

File: test/integration/util/util.go
Comment (pohly) [reply test/integration/util/util.go:132]: > how do we know what's happened?

It doesn't matter whether the context got cancelled or timed out. In both cases these functions should return. At best it's a debug message at high verbosity, not an error.

> Instead, we should probably be canceling the event handlers when the context is cancelled.

If we don't cancel the `ctx` that the fake controller is using, then there's no guarantee that the code that it is running terminates in a timely manner.

File: test/integration/scheduler_perf/util.go
Comment (pohly) [reply test/integration/scheduler_perf/util.go:130]: They are needed because tests block until all objects are truly gone before the next test starts.

Even if they are not strictly needed in performance tests (because there everything gets reset between tests, including etcd), IMHO it makes them a bit more realistic. But I can also skip them when running performance tests.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go]: Each perf test starts with a clean slate (new etcd, apiserver, scheduler). Therefore cleaning up *after* the tests is unnecessary and just slows down running them. Starting etcd and apiserver again then takes time, but that's justified because of the "clean slate".

With integration tests it's the opposite: the cost of starting etcd and apiserver gets amortized over multiple tests, but then each test has to clean up properly. That's okay because the workloads selected for integration testing tend to create less objects.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go]: All tests, performance and integration, start with no pods left over by previous tests.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (alculquicondor) [reply test/integration/scheduler_perf/scheduler_perf_test.go:35]: For E2E, I would agree, since we already use gomega intensively.

@aojea what is your take on gomega in integration tests vs the `wait` package?

> `util/wait` is awful for use in tests

Can we make it better?

Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:35]: `util/wait` is used in production code. Adding things to it that are only needed for testing (like remembering past state to produce better error messages in case of a timeout) doesn't make sense. The API also would have to be changed, because a callback that just returns "bool, error" is not informative enough (why not ready?).

We would need two versions of it, and then we are back at "let's reimplement gomega.Eventually".

File: test/integration/scheduler_perf/util.go
Comment (alculquicondor) [reply test/integration/scheduler_perf/util.go:130]: I'm indifferent, but please update the name of the function and/or comment to reflect what's happening inside the function and why.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (alculquicondor) [inline test/integration/scheduler_perf/scheduler_perf_test.go:943]: I don't see why we need to compare `objects` here.

This is a cleanup function. The pair `bool, error` from `wait` seems enough.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (chendave) [inline test/integration/scheduler_perf/scheduler_perf_test.go:889]: pre-existing, but do we need to have some prefix for this metric? such as `scheduler_` or `scheduler_perf_` ?

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (chendave) [reply test/integration/scheduler_perf/scheduler_perf_test.go]: Ack, thanks for the explanation!

**Review by chendave — COMMENTED** [review]
except some pending comments need to response, this change mostly LGTM.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (kerthcet) [reply test/integration/scheduler_perf/scheduler_perf_test.go:934]: Agree to export the cleanup as a function which makes it more readable. The runWorkload is too long now ...

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (kerthcet) [reply test/integration/scheduler_perf/scheduler_perf_test.go:945]: Benefit to debugging then.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (aojea) [reply test/integration/scheduler_perf/scheduler_perf_test.go:35]: > @aojea what is your take on gomega in integration tests vs the wait package?

Since last year I started to analyze this topic from a more objective angle and I realized that my bias to gomega was not to the Eventually method, was because of the BDD DSL patterns, that is always common on development that differentiate between unit test and integration/e2e test. Typically on companies that make the division between SW and QA engineers. I really don't like that division and I'm totally against it, a developer should cover the end to end, the feature and all the testing ( unit , integration and e2e test, ...), we also assume that in our KEPs

I really think we should avoid a BDD DSL pattern in kubernetes because of the reasons explained in the golang FAQ https://go.dev/doc/faq, it causes friction and I can't see any benefit , it is harder to debug also ...

> A related point is that testing frameworks tend to develop into mini-languages of their own, with conditionals and controls and printing mechanisms, but Go already has all those capabilities; why recreate them? We'd rather write tests in Go; it's one fewer language to learn and the approach keeps the tests straightforward and easy to understand.

So good so far, sorry for the long history, we come to a polling function for asynchronous assertion, that is `gomega.Eventually` vs `wait.Poll`.
I personally never find any issue using `wait.PollImmediate` but I can not generalize my opinion, Patrick highlighted some of the benefits of using `gomega.Eventually`in https://github.com/kubernetes/contributor-site/blob/bb1c87d320aacacc6ec5dd23d312fc59f93e3c6d/content/en/blog/2023/e2e-testing-best-practices.md#polling-and-timeouts , so my conclusion is that is a code maintainers choice to decide which function to use for an asynchronous checks and live with the pros and cons of their decision, at the end, once the code is merged, the persons responsible to maintain the code have to live with that decision.

**Review by kerthcet — COMMENTED** [review]
Looks good to me generally.

File: test/integration/util/util.go
Comment (pohly) [reply test/integration/util/util.go:132]: Stopping the informer factory when the context is canceled is [already done](https://github.com/kubernetes/kubernetes/blob/9e0569f2ed3934060fabe51be4e15232bbea3877/test/integration/util/util.go#L100).

But even with that it is possible that a sync is running or gets started when the context is canceled, which leads to useless "error while getting" error logs during shutdown.

Looking at this again, I realized that my if check wasn't in quite the right place: it must come after an API call failed, otherwise there's a race. Fixed.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:934]: > Not sure if we should wait until all pods to be deleted, is there any additional benefit from that?

It's required because different workloads use the same namespace. The test has wait until everything that might block deleting the namespace and the namespace itself are truly gone. If that fails, it should be a failure in the workload that failed to clean up, not a failure in the next workload which cannot create the namespace. Therefore we need to wait.

> Agree to export the cleanup as a function which makes it more readable. 

Done.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go]: Added.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:699]: I'm still confused...

`TestScheduling` already reuses the same apiserver and scheduler as much as possible. `BenchmarkPerfScheduling` cannot reuse them because etcd gets wiped by restarting it.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:786]: Perhaps, but it doesn't matter that much (it's usually not running that long, and `go test` itself gets called with a 10 minute timeout), so I'd prefer to not add further parameters and if checks.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:823]: Thanks, added.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:889]: I prefer to keep the name short myself, but we can discuss this further in https://github.com/kubernetes/kubernetes/issues/116204

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:943]: I added some more inline comments about the motivation for this code.

File: test/integration/scheduler_perf/util.go
Comment (pohly) [reply test/integration/scheduler_perf/util.go:130]: I renamed "setup scheduler" to "setup cluster".

**Comment by pohly** [comment]
I've gone through the review feedback and [tried to address everything](https://github.com/kubernetes/kubernetes/compare/62844bc4c1171b584d1032aa760ae78cbaca2749..0358fdb2d5060316f102f2733b324db4d7a9d65b). For now I continue to use gomega.Eventually.

@alculquicondor: if you still prefer `wait.Until`, then I'll switch to that.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (kerthcet) [reply test/integration/scheduler_perf/scheduler_perf_test.go]: We'll create the api-server inside the `mustSetupScheduler`, but actually we can preserve the api-server and re-create the scheduler for each case, the same as we did in https://github.com/kubernetes/kubernetes/pull/116634

**Comment by kerthcet** [comment]
Addressed all my concerns, so 
/approve

**Comment by kerthcet** [comment]
/hold
for @alculquicondor

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go]: @kerthcet: are you talking about the integration testing?

Remember that `mustSetupScheduler` only gets called once per unique configuration. Currently there are two of those, with and without the DRA feature gate. We cannot preserve the api-server when switching from one to the other, because the feature gate also affects it.

File: test/integration/scheduler_perf/util.go
Comment (alculquicondor) [inline test/integration/scheduler_perf/util.go:83]: It still says Scheduler here.
Also please update the comment

File: test/integration/util/util.go
Comment (alculquicondor) [inline test/integration/util/util.go:134]: This should avoid any races:

```suggestion
			if err != nil && errors.Is(err, context.Canceled) {
```

File: test/integration/util/util.go
Comment (alculquicondor) [inline test/integration/util/util.go:147]: same here

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (alculquicondor) [reply test/integration/scheduler_perf/scheduler_perf_test.go:789]: why not just `context.WithCancel` then?

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (alculquicondor) [inline test/integration/scheduler_perf/scheduler_perf_test.go:1217]: Can we use a single DeleteCollection? IIUC, it also accepts DeleteOptions

File: test/integration/scheduler_perf/util.go
Comment (pohly) [reply test/integration/scheduler_perf/util.go:83]: Fixed.

File: test/integration/util/util.go
Comment (pohly) [reply test/integration/util/util.go:134]: We have to return when there was an error, otherwise the code below would use a nil `pvc`. What needs to be fixed is the check whether that error should be logger. I had the logic backwards. It now is:

```
			if err != nil {
				// Note that the error can be anything, because components like
				// apiserver are also shutting down at the same time, but this
				// check is conservative and only ignores the "context canceled"
				// error while shutting down.
				if ctx.Err() == nil || !errors.Is(err, context.Canceled) {
					klog.Errorf("error while getting %v/%v: %v", claimRef.Namespace, claimRef.Name, err)
				}
				return
			}
```

Comment (alculquicondor) [reply test/integration/util/util.go:134]: sg

**Comment by kerthcet** [comment]
Seems forgot to push the commit...

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:1217]: Good idea, changed.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:1223]: Done.

File: test/integration/scheduler_perf/scheduler_perf_test.go
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:789]: Works for me - changed.

**Comment by pohly** [comment]
> Seems forgot to push the commit...

I didn't quite finish yesterday after leaving my initial replies. I think I addressed everything now and pushed: https://github.com/kubernetes/kubernetes/compare/2ff7891706089c1a3ae58b6cf6cd116b8e462dab..0d41d509d2d96ccc3473924cb4e1b8e1b3e4c170

**Review by alculquicondor — COMMENTED** [review]
/lgtm

**Pull request merged**
